OrbitLinuxTracing/TracerThread.cpp

```diff
@@ -35,64 +43,65 @@ void TracerThread::Run(
       std::make_unique<UprobesUnwindingVisitor>(ReadMaps(pid_));
   uprobes_unwinding_visitor->SetListener(listener_);
   // Switch between PerfEventProcessor and PerfEventProcessor2 here.
-  // PerfEventProcessor2 is supposedly faster but assumes that events from
-  // the same perf_event_open ring buffer are already sorted.
+  // PerfEventProcessor2 is supposedly faster but assumes that events from the
+  // same perf_event_open ring buffer are already sorted.
   uprobes_event_processor_ = std::make_shared<PerfEventProcessor2>(
       std::move(uprobes_unwinding_visitor));
 
   if (trace_instrumented_functions_) {
     for (const auto& function : instrumented_functions_) {
-      for (int32_t cpu = 0; cpu < num_cpus; cpu++) {
+      for (int32_t cpu : cpuset_cpus) {
         int uprobes_fd = uprobes_stack_event_open(
             function.BinaryPath().c_str(), function.FileOffset(), -1, cpu);
         PerfEventRingBuffer uprobes_ring_buffer{uprobes_fd,
                                                 BIG_RING_BUFFER_SIZE_KB};
         if (uprobes_ring_buffer.IsOpen()) {
-          fds_to_ring_buffer_.emplace(uprobes_fd,
-                                      std::move(uprobes_ring_buffer));
+          tracing_fds_.push_back(uprobes_fd);
+          ring_buffers_.push_back(std::move(uprobes_ring_buffer));
           uprobes_fds_to_function_.emplace(uprobes_fd, &function);
         }
 
         int uretprobes_fd = uretprobes_event_open(
             function.BinaryPath().c_str(), function.FileOffset(), -1, cpu);
+        tracing_fds_.push_back(uretprobes_fd);
 
         // Redirect uretprobes to the uprobes ring buffer to reduce number of
         // ring buffers and to coalesce closely related events.
         perf_event_redirect(uretprobes_fd, uprobes_fd);
-
-        // Track uretprobes fds separately as they map to a uprobes ring buffer.
-        uretprobes_fds_.push_back(uretprobes_fd);
       }
     }
   }
 
-  // TODO(b/148209993): Consider sampling based on CPU and filter by pid.
-  for (pid_t tid : ListThreads(pid_)) {
-    // Keep threads in sync.
-    listener_->OnTid(tid);
+  for (int32_t cpu : cpuset_cpus) {
+    int mmap_task_fd = mmap_task_event_open(-1, cpu);
+    PerfEventRingBuffer mmap_task_ring_buffer{mmap_task_fd,
+                                              SMALL_RING_BUFFER_SIZE_KB};
+    if (mmap_task_ring_buffer.IsOpen()) {
+      tracing_fds_.push_back(mmap_task_fd);
+      ring_buffers_.push_back(std::move(mmap_task_ring_buffer));
+    }
+  }
 
-    if (trace_callstacks_) {
-      int sampling_fd =
-          sample_mmap_task_event_open(sampling_period_ns_, tid, -1);
+  if (trace_callstacks_) {
+    for (int32_t cpu : cpuset_cpus) {
```

Comment:
Wouldn't it make sense to also do time-based sampling for all cores?

Reply:
With dynamic instrumentation, it's essentially a restriction imposed by the UI: A user can only pick a single process and then pick functions from one of the modules in that function. Hence it can never happen that you hit an event on a different process. Sampling is different and it might make sense to sample other processes, but you would then get callstacks in the UI that don't belong to the process the user intends to look at. I don't have a strong opinion on this, we can keep it as-is, and if we find a reason to add sampling across all cores, we can do that later.

Reply:
At the moment it would be very confusing to show callstacks from several processes. The user needs to select a process to sample in the UI anyway (so callstacks from other processes would be unexpected), and there are no process names in the timeline.

Reply:
Like I said, we can keep as-is and decide later if this is a feature we want. I think can make sense, but needs some thinking.

Reply:
Also I realized that sampling multiple processes would require more work because we need /proc/pid/maps for unwinding. In other words, more bookkeeping than it's currently done. So let's give this more thought first.